R/simulate_susceptibles.R

```diff
@@ -0,0 +1,165 @@
+                b = susc)
+            }
+
+    } else if (offspring  == "nbinom") {
+
+        if (missing(disp_offspring) | disp_offspring <= 1) { ## dispersion index
+            stop("Offspring distribution 'nbinom' requires argument
+                disp_offspring > 1. Use 'pois' if there is no overdispersion.")
+        }
+
+        offspring_fun <- function(n, susc) {
+            ## get distribution params from mean and dispersion
+            ## see ?rnbinom for parameter definition
+            new_mn <- mn_offspring * susc / pop ##apply susceptibility
+            size <- new_mn / (disp_offspring - 1)
+
+            ## using a right truncated nbinom distribution
+            ## to avoid more cases than susceptibles
+            truncdist::rtrunc(
+                n,
+                spec = "nbinom",
+                a = 0,
+                b = susc,
+                mu = new_mn,
+                size = size)
+        }
+    }
+
+    ## initializations
+    tdf <- data.frame(
+        id = 1L,
+        ancestor = NA_integer_,
+        generation = 1L,
+        time = t0,
+        offspring_generated = FALSE
+    )
+
+    susc <- pop - initial_immune - 1L
+    t <- t0
+
+    ## continue if any unsimulated has t <= tf
+    ## AND there is still susceptibles left
+    while (
+        any(tdf$time[!tdf$offspring_generated] <= tf) &
+        susc > 0
+        ) {
+
+        ## select from which case to generate offspring
+        t <- min(tdf$time[!tdf$offspring_generated]) #lowest unsimulated t
+
+        ## index of the first in df with t, extract vars
+        idx <- which(tdf$time == t & !tdf$offspring_generated)[1]
+        id_parent <- tdf$id[idx]
+        t_parent <- tdf$time[idx]
+        gen_parent <- tdf$generation[idx]
+
+        ## generate it
+        current_max_id <- max(tdf$id)
+        n_offspring <- offspring_fun(1, susc)
+
+        if (n_offspring %% 1 > 0) {
+            stop("Offspring distribution must return integers")
+        }
+
+        ## mark as done
+        tdf$offspring_generated[idx] <- TRUE
+
+        ## add to df
+        if (n_offspring > 0) {
+            ## draw times
+            new_times <- serial(n_offspring)
+
+            if (any(new_times < 0)) {
+                stop("Serial interval must be >= 0.")
+            }
+
+            new_df <- data.frame(
+                id = current_max_id + seq_len(n_offspring),
+                time = new_times + t_parent,
+                ancestor = id_parent,
+                generation = gen_parent + 1L,
+                offspring_generated = FALSE
+            )
+
+            ## add new cases to tdf
+            tdf <- rbind(tdf, new_df)
+        }
+
+        ## adjust susceptibles
+        susc <- susc - n_offspring
+    }
+
+    ## remove cases with time > tf that could
+    ## have been generated in the last generation
+    tdf2 <- tdf[tdf$time <= tf, ]
+
+    ## sort output and remove columns not needed
+    tdf <- dplyr::arrange(tdf, time, id)
```

Comment:
This adds the `dplyr` dependency just to sort a data frame. Is `dplyr::arrange` faster than `base::order`? If not we could use 
```r
tdf <- tdf[order(tdf$time, tdf$id), ]
```